tensorflow_recommenders_addons/dynamic_embedding/python/ops/restrict_policies.py

```diff
@@ -0,0 +1,364 @@
+    Define the rule to constrain the size of the target variable by eliminating
+    the oldest k features, and number of `residue` feature will be left.
+
+    Args:
+      residue: int. Number of remained keys after restriction.
+      **kwargs: (Optional) reserved keyword arguments.
+        trigger: int. The triggered threshold to execute restriction. Default
+          equals to `residue`.
+
+    Returns:
+      An operation to do restriction.
+    """
+    if not isinstance(residue, int):
+      raise TypeError('residue should be integer.')
+    self._residue = residue
+    trigger = kwargs.get('trigger', residue)
+    if not isinstance(trigger, int):
+      raise TypeError('trigger should be integer.')
+
+    is_overflow = math_ops.greater(self.var.size(), trigger)
+    return control_flow_ops.cond(is_overflow, self._cond_apply_restriction,
+                                 self._cond_no_op)
+
+  def _cond_apply_restriction(self):
+    restrict_var_ops, restrict_status_ops, restrict_slot_ops = [], [], []
+    for i, dev in enumerate(self.tstp_var.devices):
+      with ops.device(dev):
+        partial_keys, partial_tstp = self.tstp_var.tables[i].export()
+        partial_residue = int(self._residue / self.tstp_var.shard_num)
+        partial_tstp = array_ops.reshape(partial_tstp, (-1,))
+        first_dim = array_ops.shape(partial_tstp)[0]
+
+        k_on_top = math_ops.cast(first_dim - partial_residue,
+                                 dtype=dtypes.int32)
+        k_on_top = math_ops.maximum(k_on_top, 0)
+        _, removed_key_indices = nn_ops.top_k(-partial_tstp,
+                                              k_on_top,
+                                              sorted=False)
+        removed_keys = array_ops.gather(partial_keys, removed_key_indices)
+        restrict_var_ops.append(self.var.tables[i].remove(removed_keys))
+        restrict_status_ops.append(self.tstp_var.tables[i].remove(removed_keys))
+        for slot_param in self.params_in_slots:
+          restrict_slot_ops.append(slot_param.tables[i].remove(removed_keys))
+    return control_flow_ops.group(restrict_var_ops, restrict_status_ops,
+                                  restrict_slot_ops)
+
+  def _cond_no_op(self):
+    return control_flow_ops.no_op()
+
+  @property
+  def status(self):
+    return self.tstp_var
+
+
+class FrequencyRestrictPolicy(RestrictPolicy):
+  """
+  A derived policy to eliminate features in variable follow the
+  `lowest-occurrence-out-first` rule.
+  """
+
+  def __init__(self, var):
+    """
+    Create a FrequencyRestrictPolicy. A frequency status sparse variable is
+    created. The frequency status has same key_dtype as the target variable
+    and value_dtype in `int32`, which indicates the times of the feature
+    occurrence.
+
+    Args:
+      var: A `dynamic_embedding.Variable` object to be restricted.
+    """
+    super(FrequencyRestrictPolicy, self).__init__(var)
+    self.default_count = constant_op.constant(0, dtypes.int32)
```

Comment:
The `init_count` is just for internal use, and the frequency is elarorated in class annotation. So it might be unessesary to comment it. And further feature like *customized init_count* could be introduced.